### DISCUSSION TITLE
What are best practices for updating the cache using updateQuery?

### DISCUSSION THREAD

**Discussion by narkeeso** [urql-graphql/urql#2279] [Q&A] [answered]
The below code is a non-working example:

```
// non-working cacheExchange example
updates: {
  Mutation: {
    insert_group(result, args, cache, info) {
      if (query === "create_classroom") {
        cache.updateQuery(
          { query: QUERY, variables: { arg1: "", arg2: "" } },
          (data) => {
            result.group.unshift(result.insert_group);
            return data;
          }
        );
      }
    
      if (query === "create_family") {
        cache.updateQuery(
          { query: QUERY, variables: { arg1: "", arg2: "" } },
          (data) => {
            result.group.push(result.insert_group);
            return data;
          }
        );
      }
    },
  }
}
```

Hello,

I have a couple of questions regarding best practices for cache updating.

We are using Hasura which provides Query/Mutation resolvers that are generic and can be used for various mutations. Unlike the examples in the documentation, the mutation isn't specific like `createTodo`. Which leads me to my first question, what is the best way to send enough context so that I can run targeted `updateQuery` methods?

Some of these queries have a lot of args and the mutation might not always provide enough information to populate the variables required for the query. In Apollo my component usually had all the information it needed or I had a way to retrieve it. For the most part, I have been able to get the arguments I need to populate variables but it seems to be getting kind of messy and cacheExchange all of a sudden feels coupled to various dependencies outside of the cache exchange lifecycle. Are there any best practices for managing all this?

![image](https://user-images.githubusercontent.com/17995/153983438-cb8ca3d0-2370-4116-915d-0ebdb3912cdc.png)

The docs also mention that it's not recommended to use `result` to avoid coupling but I'm not seeing any other way to update the cache without it?

**Comment by HPieters** [comment]
Sorry not an answer, I was about to ask the same question and have the following as well. 

> "Some of these queries have a lot of args and the mutation might not always provide enough information to populate the variables required for the query."

I have resorted to using result (I found adding variables just for the updateCache and ignore them in the backend was too much). I wonder is there something possible similar to optimistic updates (see below)? 

> However, we're able to pass additional variables to our mutation, e.g. { extra }, and since $extra isn't defined it will be filtered once the mutation is sent to the API. An optimistic mutation however will still be able to access this variable.

**Comment by JoviDeCroock** [comment, marked as answer]
> what is the best way to send enough context so that I can run targeted updateQuery methods?

You can opt to send extra `variables` in your `mutation` that won't be sent to your server but will be available on the fourth argument of your updater function on `info.variables`. Generally I would say to avoid generic mutations as everything should be derivable from `result.yourField.__typename`.

> The docs also mention that it's not recommended to use result to avoid coupling but I'm not seeing any other way to update the cache without it?

The quote is telling you to only look at the current field that is being mutated rather than the full result, this because you can add as much operations to one query as you like.

**Reply by romain-trotard** [reply]
Hello @JoviDeCroock 
I am really interested by passing extra variables to mutation. There is a part in the [documentation](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#variables-for-optimistic-updates) but no example.
Unfortunately I don't know how to make it work. I made a [codesandbox](https://codesandbox.io/s/nostalgic-curran-436ms3?file=/src/index.js). In the `Todo` component I pass an extra variable named `other` and in the `index.js`, I log `info.variables`. 
Is it possible to explain me how to do this please?
Thanks

**Reply by adamf321** [reply]
+1 for an example here please. I am also trying to figure out how this works. I tried adding an extra variable to my mutation, but it is being sent to the graphql server and I get an error: `Field "tmpId" is not defined by type "CreateOkrInput".`

**Reply by JoviDeCroock** [reply]
The codesandbox of @romain-trotard is correct, we had a bug which is solved in the canary release and will be fixed in stable when we release #3040

**Reply by adamf321** [reply]
Thank you both. For anyone else reading - I got it working like this:

```
const [, executeMutation] = useMutation(CreateOkrDocument)

executeMutation({
   data: {
      ...       
   },
   extraField: 'abc',
})
```

This allowed me to access `info.variables.extraField` in the optimistic cache exchange. And it works with urql v2.0.4.
